modules/geochemistry/src/utils/EquilibriumGeochemicalSystem.C

```diff
@@ -0,0 +1,1310 @@
+(activity_coefficient * basis molality)^stoi, so that the derivative is quite simple.  Eg, we
+don't have to worry about derivatives with respect to fixed-activity things, or fixed fugacity.
+Also, note that the constructor and the various "set" methods of this class enforce molality > 0,
+so there are no division-by-zero problems.  Also, note that we never compute derivatives of the
+activity coefficients, or the activity of water with respect to the molalities, as they are
+assumed to be quite small.  Finally, the surface_pot_expr will also always be positive.
+  */
+  // correctly size and zero
+  jac.resize(_num_in_algebraic_system, _num_in_algebraic_system);
+  const Real nw = _basis_molality[0];
+
+  // jac(a, b) = d(R_a) / d(m_b), where a corresponds to a molality
+  for (unsigned a = 0; a < _num_basis_in_algebraic_system; ++a)
+  {
+    const unsigned basis_of_a = _basis_index[a];
+    for (unsigned b = 0; b < _num_basis_in_algebraic_system; ++b)
+    {
+      const unsigned basis_of_b = _basis_index[b];
+      if (basis_of_b == 0) // special cases: mass of solvent water is an unknown
+      {
+        if (basis_of_a != _charge_balance_basis_index)
+          jac(a, 0) += (res(a) + _bulk_moles[basis_of_a]) / nw;
+        else
+        {
+          Real extra_term = 0.0;
+          for (unsigned i = 0; i < _num_basis; ++i)
+          {
+            if (i == _charge_balance_basis_index)
+              continue;
+            else if (_mgd.basis_species_charge[i] ==
+                     0.0) // certainly includes water, minerals and gases
+              continue;
+            else if (_in_algebraic_system[i]) // bulk moles must have been provided by user
+              extra_term += _mgd.basis_species_charge[i] * _bulk_moles[i] /
+                            _mgd.basis_species_charge[basis_of_a];
+            // note that if species i has ACTIVITY or FREE_MOLALITY set then its bulk moles is
+            // linear in nw so it should not appear in extra_term
+          }
+          jac(a, 0) += (res(a) - extra_term) / nw;
+        }
+      }
+      else
+      {
+        if (a == b)
+          jac(a, b) += nw;
+        for (unsigned eqm_j = 0; eqm_j < _num_eqm; ++eqm_j)
+        {
+          jac(a, b) += nw * _mgd.eqm_stoichiometry(eqm_j, basis_of_a) * _eqm_molality[eqm_j] *
+                       _mgd.eqm_stoichiometry(eqm_j, basis_of_b) / _basis_molality[basis_of_b];
+        }
+        if (basis_of_a == _charge_balance_basis_index)
+        {
+          // additional terms from the charge-balance additions
+          for (unsigned i = 0; i < _num_basis; ++i)
+          {
+            if (i == _charge_balance_basis_index)
+              continue;
+            else if (_mgd.basis_species_charge[i] ==
+                     0.0) // certainly includes water, minerals and gases
+              continue;
+            else if (_constraint_meaning[i] == ConstraintMeaningEnum::ACTIVITY ||
```

Comment:
Hard to tell but should this be else instead of else if? It might be just how it looks on GitHub

Reply:
Again, i *think* this should be `else if`